feat: Support business code capture interruption during exit

marmot-z opened the pull request:
Fixes [2473](https://github.com/apache/shardingsphere-elasticjob/issues/2473).

Changes proposed in this pull request:
- Support business code capture interruption and graceful exit while invoke JobScheduler#shutdown method

marmot-z committed:
feat: Support business code capture interruption during exit

marmot-z committed:
fix: maintain variables correctly at the end

marmot-z commented:
Dear project maintainers, could you please review the code and merge it for me?

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,47 @@ void testShardingStatisticsAPI() {
+    @Test
+    void testGivenOneMoreShardingCountWhenShutdownThenTaskCanCaptureInterruptedException() throws Exception {
+        testCaptureInterruptedException(2);
+    }
```
- I will say that I'm confused as to why the unit tests for this change are placed in a separate integration test module.

marmot-z replied:
Because I found that the logic of using one shardingTotalCount and multiple shardingTotalCounts is different. When there is only one shardingTotalCount, the task always runs on the Quartz thread. When there are multiple shardingTotalCounts, the task runs on the ElasticJobExecutor thread. So I write two verify cases. 
<img width="1298" height="557" alt="image" src="https://github.com/user-attachments/assets/2e2f10ca-a1a6-4049-9736-d8d7369cbef5" />

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,47 @@ void testShardingStatisticsAPI() {
+    private void testCaptureInterruptedException(int shardingTotalCount) throws Exception {
+        String jobName = "testTaskCaptureInterruptedTask";
+        final AtomicBoolean captured = new AtomicBoolean(false);
+        // start in 2 seconds
```
- I have the impression that ShardingSphere's code constraints don't like inline comments, as Javadoc doesn't display them. Can you change the comments to Javadoc style, or remove the inline comments?

marmot-z replied:
Ok, I will fix it as soon as possible.

linghengqian reviewed (commented):
The code style is handled in a separate Maven profile. The PR did not pass the code formatting validation. You can execute first `./mvnw spotless:apply -Pcheck -T1C`. Then execute `./mvnw checkstyle:check -Pcheck -T1C` to manually adjust to repair the CI.

marmot-z commented on kernel/src/main/java/org/apache/shardingsphere/elasticjob/kernel/internal/schedule/LiteJob.java:
```diff
@@ -19,20 +19,37 @@
+            jobExecutor.execute();
+        } finally {
+            currentThread = null;
+        }
```
Ok, I will fix it as soon as possible.

marmot-z commented:
> The code style is handled in a separate Maven profile. The PR did not pass the code formatting validation. You can execute first `./mvnw spotless:apply -Pcheck -T1C`. Then execute `./mvnw checkstyle:check -Pcheck -T1C` to manually adjust to repair the CI.

OK, I got that.

marmot-z commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,47 @@ void testShardingStatisticsAPI() {
+                captured.set(true);
+                Thread.currentThread().interrupt();
+            }
+        };
```
Ok, I will fix it as soon as possible.

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,47 @@ void testShardingStatisticsAPI() {
+                        .cron(cronExpression)
+                        .build());
+        job.schedule();
+        Thread.sleep(4_000);
```
I'll say I'm confused as to why `Thread#sleep` is necessary. What happens if we use `Awaitility.await().pollDelay(4L, TimeUnit.SECONDS).until(() -> true);`? https://github.com/apache/shardingsphere/issues/24106 is introducing stricter code constraints to avoid `Thread#sleep`.

marmot-z replied:
wow, that's cool. I didn't know it before, but now I get it. I will fix my test code later.

marmot-z committed:
fix: fix known issue and code styles

marmot-z commented:
I have fixed the above questions, PTAL  @linghengqian

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,44 @@ void testShardingStatisticsAPI() {
+        job.schedule();
+        Awaitility.await().pollDelay(4L, TimeUnit.SECONDS).until(() -> true);
+        job.shutdown();
+        Awaitility.await().pollDelay(1L, TimeUnit.SECONDS).untilAsserted(() -> assertThat(captured.get(), is(true)));
```
- If you're just verifying the asynchronous result of `job.shutdown();`, why use `pollDelay` to slow down the execution of the unit test? Why not use `Awaitility.await().atMost(Duration.ofSeconds(30L)).ignoreExceptions().until(() -> captured.get());`?
- Or, do you need additional assertions for exceptions? I don't see any such assertions in your commit.

marmot-z replied:
This requires four steps:
1. Wait 4 seconds for the job to actually run (because the task is set to start within two seconds).
2. Shut down the elastic-job.
3. Wait 1 second for the elastic-job to execute its exit logic.
4. Check that the variable status matches expectations.

linghengqian replied:
From this perspective, there's actually no need to wait for 1 second for the corresponding variable to be set to `true`. Am I right? 

Or do you actually need an `AtomicReference<String>` with three possible values ​​as `captured` because your unit test is designed to handle three situations?

marmot-z replied:
I think declaring `captured` as an AtomicBoolean field is sufficient. As long as it becomes true, it means that the business code can capture the interrupt, which is in line with expectations. 

However, I am still not sure what the problem is here. Is it that the `Awaitility` API is used incorrectly, the waiting time is set too long, or the logic of this test code is not clear? 
1. If the `Awaitility` API is used incorrectly, please give me some tips. 
2. If the waiting time is set too long, I will try to shorten it. 
3. If the test code logic is unclear, please point out the unclear part.

linghengqian replied:
Let me explain my point. I don't believe `job.shutdown();` needs to wait for a specific amount of time, such as `1` second. As long as `captured` is `true`, the changes in the current PR are valid.

From this perspective, there's no need to use `Awaitility.await().pollDelay()` to simulate the behavior of `Thread#sleep`. In this case, `Awaitility.await().atMost(Duration.ofSeconds(5L)).ignoreExceptions().until(() -> captured.get());` is sufficient, or perhaps a longer timeout than `5` seconds is sufficient.

linghengqian replied:
Oh, I just saw your changes in another commit. I think the details can be discussed in another PR. This is no longer an issue.

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,44 @@ void testShardingStatisticsAPI() {
+                    System.out.println("Running...");
+                    Thread.sleep(100);
+                }
+            } catch (InterruptedException e) {
```
- How about changing it to `} catch (final InterruptedException ex) {` ?
- I'm surprised that Elasticjob's checkstyle configuration doesn't check for `final` keywords and variable names like ShardingSphere does.

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,44 @@ void testShardingStatisticsAPI() {
+
+    private void testCaptureInterruptedException(int shardingTotalCount) throws Exception {
+        String jobName = "testTaskCaptureInterruptedTask";
+        final AtomicBoolean captured = new AtomicBoolean(false);
```
- Why do local variables within functions need to be declared `final`? Is this a suggestion from GitHub Copilot?

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,44 @@ void testShardingStatisticsAPI() {
+    @Test
+    void testGivenOneMoreShardingCountWhenShutdownThenTaskCanCaptureInterruptedException() throws Exception {
+        testCaptureInterruptedException(2);
+    }
```
- Why can't these two unit tests be merged into one? Will merging them into the same unit test cause conflicts? Elasticjob unit tests are currently executed serially. What am I missing?
```java
{
testCaptureInterruptedException(1);
testCaptureInterruptedException(2);
}
```

marmot-z replied:
I have fixed the above questions, PTAL @linghengqian

linghengqian reviewed (commented):
- It looks like some Mockito-related assertions are failing. If you're using PowerShell 7, you can run `./mvnw --batch-mode --no-transfer-progress '-Dmaven.javadoc.skip=true' clean install -T1C > ./log.txt` to see which unit tests are failing.

marmot-z committed:
fix: fix known issue

marmot-z committed:
fix: fix related issue and test code

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,40 @@ void testShardingStatisticsAPI() {
+        testCaptureInterruptedException(2);
+    }
+
+    private void testCaptureInterruptedException(int shardingTotalCount) throws Exception {
```
Checkstyle's CI is still failing, variables in method signatures need to be explicitly declared `final`.

linghengqian commented on kernel/src/main/java/org/apache/shardingsphere/elasticjob/kernel/internal/schedule/LiteJob.java:
```diff
@@ -19,20 +19,37 @@
     public void execute(final JobExecutionContext context) {
-        jobExecutor.execute();
+        currentThread = Thread.currentThread();
+
```
This blank line should be removed.

marmot-z replied:
I have fixed the above questions, PTAL @linghengqian

marmot-z committed:
refactor: refactor test code

linghengqian reviewed (commented):
You need to run `./mvnw spotless:apply -Pcheck -T1C` to format the code.

linghengqian committed:
Bump Curator to `5.9.0` to fix incompatible changes around Zookeeper Client `3.9.3`

marmot-z commented:
> You need to run `./mvnw spotless:apply -Pcheck -T1C` to format the code.

I ran it locally but nothing changed.
<img width="658" height="318" alt="image" src="https://github.com/user-attachments/assets/652568f7-eb58-4e98-995d-5f10a5db84a9" />

linghengqian reviewed (commented):
The failure of unit tests under GraalVM Native Image is not related to the current PR, I have opened https://github.com/apache/shardingsphere-elasticjob/pull/2476 .

marmot-z commented:
> The failure of unit tests under GraalVM Native Image is not related to the current PR, I have opened #2476 .

What else do you need me to do?

linghengqian commented:
> > The failure of unit tests under GraalVM Native Image is not related to the current PR, I have opened #2476 .
> 
> What else do you need me to do?

- There's nothing you need to do right now. If you don't mind me reviewing this PR tomorrow, I need to fix the failing unit tests on the master branch first.🤣 
- No one had told me before that the nativeTest branch of the elasticjob master branch was broken because the relevant unit tests were not always triggered by CI and few developers used GraalVM. This is why I previously recommended that you put your unit tests in a folder like `bootstrap/src/test/java/org/apache/shardingsphere/elasticjob/bootstrap`.
- I also found that elasticjob is currently not compilable with JDK 24, but this does not affect the current PR. I will open an issue later to investigate what happened.

marmot-z commented:
> > > The failure of unit tests under GraalVM Native Image is not related to the current PR, I have opened #2476 .
> > 
> > 
> > What else do you need me to do?
> 
> * There's nothing you need to do right now. If you don't mind me reviewing this PR tomorrow, I need to fix the failing unit tests on the master branch first.🤣
> * No one had told me before that the nativeTest branch of the elasticjob master branch was broken because the relevant unit tests were not always triggered by CI and few developers used GraalVM. This is why I previously recommended that you put your unit tests in a folder like `bootstrap/src/test/java/org/apache/shardingsphere/elasticjob/bootstrap`.
> * I also found that elasticjob is currently not compilable with JDK 24, but this does not affect the current PR. I will open an issue later to investigate what happened.

Ok, Thank you for your patient review and selfless dedication 😄 👍

linghengqian reviewed (commented):
You can try rebasing to the master branch and then executing `./mvnw spotless:apply -Pcheck -T1C`. After checking out your branch, I saw that `spotless` additionally handles the formatting of blank lines.

marmot-z commented:
> You can try rebasing to the master branch and then executing `./mvnw spotless:apply -Pcheck -T1C`. After checking out your branch, I saw that `spotless` additionally handles the formatting of blank lines.

I ran it locally, but it was always noting no change. Maybe the runtime environment of my local is different from the CI runtime environment. My local runtime environment is below:
```
> ~/shardingsphere-elasticjob[feat-support-interrupt]  ./mvnw spotless:apply -Pcheck -T1C -X
Apache Maven 3.9.8 (36645f6c9b5079805ea5009217e36f2cffd34256)
Maven home: /Users/xxx/.m2/wrapper/dists/apache-maven-3.9.8/af622e91
Java version: 1.8.0_292, vendor: AdoptOpenJDK, runtime: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre
Default locale: en_CN, platform encoding: UTF-8
OS name: "mac os x", version: "10.16", arch: "x86_64", family: "mac"
```

linghengqian commented:
The humorous thing about this question is that I've never used MacOS, so I'm not sure if there are any bugs with the spotless plugin on MacOS. 

You can rebase the master branch first, and then I'll push another commit to your branch to fix it.

marmot-z commented:
> The humorous thing about this question is that I've never used MacOS, so I'm not sure if there are any bugs with the spotless plugin on MacOS.
> 
> You can rebase the master branch first, and then I'll push another commit to your branch to fix it.

I have already rebased the master branch.

linghengqian commented:
> > The humorous thing about this question is that I've never used MacOS, so I'm not sure if there are any bugs with the spotless plugin on MacOS.
> > You can rebase the master branch first, and then I'll push another commit to your branch to fix it.
> 
> I have already rebased the master branch.

- It looks like you haven't pushed to the corresponding branch. I see that the top of https://github.com/marmot-z/shardingsphere-elasticjob/tree/feat-support-interrupt is https://github.com/marmot-z/shardingsphere-elasticjob/commit/f898217f09f23395ed3b0d4864d69f056185d9d1 , but there is no https://github.com/apache/shardingsphere-elasticjob/commit/cd7af976dc79df83fa411e332a5bfe3e5dd9f62b .

marmot-z committed:
Merge branch 'apache:master' into feat-support-interrupt

marmot-z committed:
Merge branch 'feat-support-interrupt' of github.com:marmot-z/shardingsphere-elasticjob into feat-support-interrupt

marmot-z commented:
> > > The humorous thing about this question is that I've never used MacOS, so I'm not sure if there are any bugs with the spotless plugin on MacOS.
> > > You can rebase the master branch first, and then I'll push another commit to your branch to fix it.
> > 
> > 
> > I have already rebased the master branch.
> 
> * It looks like you haven't pushed to the corresponding branch. I see that the top of https://github.com/marmot-z/shardingsphere-elasticjob/tree/feat-support-interrupt is [marmot-z@f898217](https://github.com/marmot-z/shardingsphere-elasticjob/commit/f898217f09f23395ed3b0d4864d69f056185d9d1) , but there is no [cd7af97](https://github.com/apache/shardingsphere-elasticjob/commit/cd7af976dc79df83fa411e332a5bfe3e5dd9f62b) .

Sorry, my repository is a fork repository, I should sync the fork from the main repository before rebasing the main branch. Now I have rebased the newest code into my branch.

linghengqian committed:
Apply spotless code formatting

linghengqian reviewed (approved):
- LGTM.

linghengqian commented on test/native/src/test/java/org/apache/shardingsphere/elasticjob/test/natived/it/operation/JavaTest.java:
```diff
@@ -305,4 +308,45 @@ void testShardingStatisticsAPI() {
+                running.set(true);
+                
+                while (true) {
+                    if (Thread.currentThread().isInterrupted()) {
+                        captured.set(true);
+                        running.set(false);
+                        break;
+                    }
```
- From what I can see in my testing, the unit test never actually executes the logic inside the `if` block. Does this make any sense?

marmot-z replied:
It has no sense here, you can delete it.
